pkg/controller/clusters/workers.go

```diff
@@ -0,0 +1,122 @@
+package clusters
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	kubeadmv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/clusterapi"
+	"github.com/aws/eks-anywhere/pkg/collection"
+	"github.com/aws/eks-anywhere/pkg/controller"
+	"github.com/aws/eks-anywhere/pkg/controller/serverside"
+)
+
+// Workers represents the CAPI spec for an eks-a cluster's workers.
+type Workers struct {
+	Groups []WorkerGroup
+}
+
+// objects returns a list of API objects for a collection of worker groups.
+func (w *Workers) objects() []client.Object {
+	objs := make([]client.Object, 0, len(w.Groups)*3)
+	for _, g := range w.Groups {
+		objs = append(objs, g.objects()...)
+	}
+
+	return objs
+}
+
+// WorkerGroup represents the CAPI spec for an eks-a worker group.
+type WorkerGroup struct {
+	KubeadmConfigTemplate   *kubeadmv1.KubeadmConfigTemplate
+	MachineDeployment       *clusterv1.MachineDeployment
+	ProviderMachineTemplate client.Object
+}
+
+func (g *WorkerGroup) objects() []client.Object {
+	return []client.Object{
+		g.KubeadmConfigTemplate,
+		g.MachineDeployment,
+		g.ProviderMachineTemplate,
+	}
+}
+
+// ToWorkers converts the generic clusterapi Workers definition to the concrete one defined
+// here. It's just a helper for callers generating workers spec using the clusterapi package.
+func ToWorkers[M clusterapi.Object[M]](capiWorkers *clusterapi.Workers[M]) *Workers {
+	w := &Workers{
+		Groups: make([]WorkerGroup, 0, len(capiWorkers.Groups)),
+	}
+
+	for _, g := range capiWorkers.Groups {
+		w.Groups = append(w.Groups, WorkerGroup{
+			MachineDeployment:       g.MachineDeployment,
+			KubeadmConfigTemplate:   g.KubeadmConfigTemplate,
+			ProviderMachineTemplate: g.ProviderMachineTemplate,
+		})
+	}
+
+	return w
+}
+
+// ReconcileWorkersForEKSA orchestrates the worker node reconciliation logic for a particular EKS-A cluster.
+// It takes care of applying all desired objects in the Workers spec and deleting the
+// old MachineDeployments that are not in it.
+func ReconcileWorkersForEKSA(ctx context.Context, log logr.Logger, c client.Client, cluster *anywherev1.Cluster, w *Workers) (controller.Result, error) {
```

Comment:
The generic version of `Workers` introduces complexity (bc it uses generics). That complexity has some benefits for spec generation, since it allows to have a shared implementation for all providers that still return concrete api types. However, the code in this method doesn't need to know about concrete provider types and using an interface simplifies it. I would admit I don't like having the same object defined twice, but I find this tradeoff better than dragging the generic complexity over to this package.

I could define the non generic version of Workers in the clusterapi package but I decided to start here and move it there if we see the need to use it somewhere else.

As far as moving this whole method to `clusterapi`, i don't think it's the right place for it. `clusterapi` was supposed to deal with capi object generation and api utils in general. It's not for controller specific code. We could expand the scope, but I don't see any benefit, given that `controller/clusters` already existed and its current scope is exactly that: reusable cluster controller specific code.

Reply:
Ok yeah, that's fair! I also don't love the duplication but you made some valid points. Thanks for the in-depth explanation!